src/security/srv_acl.c

```diff
@@ -499,8 +498,20 @@ ds_sec_pool_get_capabilities(uint64_t flags, d_iov_t *cred,
 
 	rc = get_sec_capas_for_token(token, ownership, acl, 0 /* no special owner perms */,
 				     pool_capas_from_perms, capas);
-	if (rc == 0)
+	if (rc == 0) {
 		filter_pool_capas_based_on_flags(flags, capas);
+		if (immutable) {
```

Comment:
simpler than this, I wonder about checking this at the handler level:
```c
if (immutable && (flags != DAOS_PC_RO)) {
        D_ERROR("caller requested RW permissions on immutable pool\n");
        return -DER_NO_PERM;
}
```
Then I don't think any change would be needed in the security capas code at all. If you make the simpler check early in the RPC handlers, only RO requests will get through to this point, so the capas won't need to be further filtered.

Reply:
Yes, I have ever implemented like that, it can work. But as former discussion, we hoped to unify the check `if (immutable && (flags != DAOS_PC_RO))` into `ds_sec_pool_can_connect()`. That will change `ds_sec_pool_can_connect()`. So either `ds_sec_pool_get_capabilities()` or `ds_sec_pool_can_connect()` needs some enhancement to support `immutable` feature. I prefer to change `ds_sec_pool_get_capabilities()`. Because if still want to implement my original idea about the security capability `ds_sec_pool_get_capabilities()` in future, then we do not need to re-change related security interfaces.

Reply:
On seeing the code, I guess the question for me is whether it is useful to unify the immutable check into the capas calculation at all. We don't even need to get the ACL/owner props to know the user has made an invalid request, asking for RW permissions to an immutable pool. It feels almost like an input consistency check, rather than a capability calculation.

Reply:
OK, I will move `immutable` check out side of security module, instead, related RPC handler (for pool connection and container open) will handle that explicitly.